contributors/design-proposals/api-machinery/customresources-versioning.md

```diff
@@ -0,0 +1,177 @@
+  // Order: The order of these versions is used to determine the order in discovery API
+  // (preferred version first).
+  // The versions in this list may not be removed if they are in 
+  // CustomResourceDefinitionStatus.StoredVersions list.
+  Versions []CustomResourceDefinitionVersion
+}
+
+// ***************
+// ** New Type **
+// ***************
+type CustomResourceDefinitionVersion {
+  // Name is the version name, e.g. "v1", “v2beta1”, etc.
+  Name string
+  // Served is a flag enabling/disabling this version from being served via REST APIs
+  Served Boolean
+  // Storage flags the release as a storage version. There can be only one version
+  // flagged as Storage.
+  Storage Boolean
+}
+```
+
+The Status object will have a list of potential stored objects. This data is necessary to do a storage migration in future (the author can choose to do the migration themselves but there is a plan [TODO(lavalamp): link?] to solve the problem of migration, potentially for both standard and custom types).
+
+```golang
+// CustomResourceDefinitionStatus indicates the state of the CustomResourceDefinition
+type CustomResourceDefinitionStatus struct {
+  ...
+
+  // StoredVersions are all versions ever marked as storage in spec. Tracking these
+  // versions allow a migration path for stored version in etcd. The field is mutable
+  // so the migration controller can first make sure a version is certified (i.e. all
+  // stored objects is that version) then remove the rest of the versions from this list.
+  // None of the versions in this list can be removed from the spec.Versions field.
+  StoredVersions []string
+}
+```
+
+Basic validations needed for the `version` field are:
+
+* `Spec.Version` field exists in `Spec.Versions` field (serve should be True for that version if there is any).
+* If there is any `Served` `Version` in list of `Versions`, then the `Spec.Version`’s version in the `Spec.Versions` list should have a set `Served` flag. This is for backward compatibility. An old controller expect that version to be served but only the whole CRD is served. CRD Registration controller should unregister a CRD with no serving version.
+* None of the `Status.StoredVersion` can be removed from `Spec.Versions` list.
+* Only one of the versions in `spec.Versions` can flag as `Storage` version.
+
+# **Declarative conversion**
+
+Most of CRD conversions can be represented by simple operations like rename or move. The proposal is to support rename only and add more operations in future based on user feedback.
+
+```golang
+type CustomResourceDefinitionVersion {
+  // Name is the version name, e.g. "v1", “v2beta1”, etc.
+  Name string
+  // Served is a flag enabling/disabling this version from being served via REST APIs
+  Served Boolean
+  // Storage flags the release as a storage version. There can be only one version
+  // flagged as Storage.
+  Storage Boolean
+  //
+  Conversion CustomResourceConversion
+}
+
+type CustomResourceConversion struct {
```

Comment:
This is more involved than the "no changes" we had talked about before.  How about we start small?

Reply:
The first PR will be no change, but for 1.11 we want at least rename. Do you have any objection to the rename design here?